### PR TITLE
Graphite output plugin: Enable secure connection

### PR DESCRIPTION
**Pull request by dheerajdwivedi**
### Proposal:
We are looking to use hosted solution for graphite.
Since the communication is over internet it needs to be secure.
Hence the support for tls over tcp is required.

### Current behavior:
Insecure communication between telegraf client and server

### Desired behavior:
Secure the communication between client and server

### Use case:
Graphite hosted solutions where data flows over internet.
To avoid middleman attack, hence secure the auth credentials and metrics data between client and server

### CODE REVIEW

**Comment by danielnelson** [comment]
#2600

**Comment by dheerajdwivedi** [comment]
False failure can someone rebuild it please.

**Comment by nhaugo** [comment]
@dheerajdwivedi can you write some tests so we can support this long term.

File: plugins/outputs/graphite/graphite.go [hunk 1]
```diff
@@ -34,6 +50,16 @@ var sampleConfig = `
   template = "host.tags.measurement.field"
   ## timeout in seconds for the write connection to graphite
   timeout = 2
+
+	## Enable secure tunnel
+	ssl_enabled = false
```
Comment (danielnelson) [inline plugins/outputs/graphite/graphite.go]: I think we don't need this option, enabling a secure connection can be based on the tlsConfig being nil or not.

File: plugins/outputs/graphite/graphite.go [hunk 2]
```diff
@@ -34,6 +50,16 @@ var sampleConfig = `
...
+
+	## Optional SSL Config
+	# ssl_ca = "/etc/telegraf/ca.pem"
+	# ssl_cert = "/etc/telegraf/cert.pem"
+	# ssl_key = "/etc/telegraf/key.pem"
+	## Use SSL but skip chain & host verification
+	# insecure_skip_verify = false
```
Comment (danielnelson) [inline plugins/outputs/graphite/graphite.go]: Indent this section to 2 spaces

File: plugins/outputs/graphite/graphite.go [hunk 3]
```diff
@@ -44,11 +70,28 @@ func (g *Graphite) Connect() error {
...
+	}
+
 	// Get Connections
 	var conns []net.Conn
 	for _, server := range g.Servers {
-		conn, err := net.DialTimeout("tcp", server, time.Duration(g.Timeout)*time.Second)
+		var conn net.Conn
+		if g.SSLEnabled {
+			// If ssl is enabled make secure tcp connection
+			conn, err = tls.Dial("tcp", server, g.tlsConfig)
```
Comment (danielnelson) [inline plugins/outputs/graphite/graphite.go]: Need to use timeout with tls connections.  Should be able to do this using tls.DialWithDialer.

I think its possible to just use a nil tls.Config to do a insecure connection, but not sure.  This way you don't need the branch.

File: plugins/outputs/graphite/graphite.go [hunk 4]
```diff
@@ -44,11 +70,28 @@ func (g *Graphite) Connect() error {
...
+		} else {
+			// Create insecure tcp connection by default
+			conn, err = net.DialTimeout("tcp", server, time.Duration(g.Timeout)*time.Second)
+		}
 		if err == nil {
+			// Append successful connections
```
Comment (danielnelson) [inline plugins/outputs/graphite/graphite.go]: remove

**Comment by danielnelson** [comment]
Thanks!  Will be included in 1.4.0.

**Comment by michael-lohier** [comment]
Hello,

 @danielnelson @dheerajdwivedi 

The current configuration to establish an TLS/SSL connection between Telegraf and Graphite is made via the output [plugin](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/graphite/README.md), setting these options in Telegraf:

```
  ## Optional TLS Config
  # tls_ca = "/etc/telegraf/ca.pem"
  # tls_cert = "/etc/telegraf/cert.pem"
  # tls_key = "/etc/telegraf/key.pem"
  ## Use TLS but skip chain & host verification
  # insecure_skip_verify = false
```
However, on Graphite side, it doesn't seem so clear how to set it up. I assume it is possible (or at least  it was at the time you added this option). Could you please tell me how was the setup with Graphite you used to test the TLS/SSL connection between Telegraf and Graphite, at the time you added these options?

Thanks a lot

**Pull request merged**